### DISCUSSION TITLE
To reintroduce or not to reintroduce IMAP capabilities?

### DISCUSSION THREAD

**Discussion by gianlucagiacometti** [postfixadmin/postfixadmin#869] [Ideas]
Hello everybody,
What about reintroducing IMAP capabilities?

According to https://php.watch/versions/8.4/imap-unbundled php_imap, will be completely over from php 8.4, and it has already been quite rightly eradicated from postfixAdmin code.
As suggested, it seems that https://github.com/Webklex/php-imap could be a good replacement.

I'm going to use it via the new post-creation hook to create sub-folders as in the old times, and it would also be useful to create tools for mail admins such as emptying the trash or spam folder, checking folders, ecc ecc.

Do you think it could be a good idea, or it does it go beyond the scope of postfixAdmin?

**Comment by gianlucagiacometti** [comment]
I just tried Webklex/php-imap and I was able to create folders with a very short function.
What I don't like is the fact that this library depends on no less that 25 other libraries, and the documentation is not complete (the subscribe function is not mentioned anywhere but in the code itself).
I spent a couple of hours looking for another imap library not relying on the php-imap extension, with no success.
So far, either using Webklex/php-imap, or building a new one from scratch. I'll keep exploring until PHP 8.4 gets into the next Ubuntu LTS.

**Comment by DavidGoodwin** [comment]
My random 2p/2c ... with the new **mailbox_postcreation_hook** you added in #864 you could decide to keep a local dependency on imap - and use the extension if you so wished .. without PostfixAdmin (the project/code) having to know or care?

The hook function could be fairly simple I guess - e.g. a bit like this pseudo-code : 

```PHP
$CONF['mailbox_postcreation_hook'] = function($param1, $param2) { 
     if(!extension_exists('imap') || !function_exists('imap_open')) {
         error_log("imap extension not available or imap_open() does not exist. Can't call mailbox_postcreation_hook");
        return false;
    }
    // connect to your imap server, do magic here??
   try { 
      // ........ 
   } catch (\Exception $e) {
         error_log("postfiadmin - error calling  mailbox_postcreation_hook" . $e->getMessage());
   }
};
```

**Reply by gianlucagiacometti** [reply]
Thank you David.
I proposed the hook exactly for that purpose, since in my servers I need to create the folders anyway.

For the moment, I re-used the old function you had in the previous versions, but, of course, I need to start thinking about getting rid of the native php imap extension (PHP 8.4 has no such extension and according to #812 it should not be used anyway). Webklex/php-imap seems fit for the purpose, though not in an ideal way for all its dependencies, and the problem of folders is anyway solved.

I then started thinking about other possible uses of an imap tool within postfixAdmin, to avoid any manual work via cli on mailboxes and domains.
Consequently, I started this discussion just to check if the developer team could be interested in this or if it's just something that I need personally.
If the developer team is interested in any use of it, I could try to develop an imap class, as simple as possible, just for the purpose.

Or I just think too much :-D
